File: src/lib/evaluations.ts
import Database from "better-sqlite3";
import { randomUUID } from "crypto";
import { execFileSync, spawn } from "child_process";
import { existsSync } from "fs";
import { homedir } from "os";
import { join } from "path";
import { getConversation } from "@/lib/claude-data";
import type {
  ConversationEvaluation,
  EvaluationPrompt,
  ProcessedConversation,
} from "@/lib/types";
import type {
  EvaluationProvider,
  EvaluationScope,
} from "@/lib/evaluation-models";

const OLTP_DB_PATH = join(
  process.cwd(),
  "public",
  "database-artifacts",
  "oltp",
  "helaicopter_oltp.sqlite"
);

export const DEFAULT_EVALUATION_PROMPT_NAME = "Default Conversation Review";
export const DEFAULT_EVALUATION_PROMPT = `Review this assistant conversation as an operator trying to improve future prompts, instructions, and conversation flow.

Focus on:
- unclear or under-specified user instructions
- avoidable tool failures and recovery quality
- places where the assistant should have clarified sooner
- bloated or distracting turns that increased cost without moving the task forward
- concrete rewrites that would make the next run cleaner

Return markdown with these sections:
## Executive Summary
## Instruction Problems
## Conversation Flow Problems
## Concrete Prompt Improvements
## Concrete Recovery Improvements
## Suggested Better Opening Prompt
## Top 3 Highest-Leverage Changes

Every recommendation should be concrete, actionable, and tied to specific message ids or tool calls when possible.`;

let databaseReady = false;

type PromptRow = {
  prompt_id: string;
  name: string;
  description: string | null;
  prompt_text: string;
  is_default: number;
  created_at: string;
  updated_at: string;
};

type EvaluationRow = {
  evaluation_id: string;
  conversation_id: string;
  prompt_id: string | null;
  provider: EvaluationProvider;
  model: string;
  status: "running" | "completed" | "failed";
  scope: EvaluationScope;
  selection_instruction: string | null;
  prompt_name: string;
  prompt_text: string;
  report_markdown: string | null;
  raw_output: string | null;
  error_message: string | null;
  command: string;
  created_at: string;
  finished_at: string | null;
  duration_ms: number | null;
};

export interface EvaluationRequest {
  projectPath: string;
  sessionId: string;
  provider: EvaluationProvider;
  model: string;
  scope: EvaluationScope;
  promptId?: string | null;
  promptName?: string | null;
  promptText: string;
  selectionInstruction?: string | null;
}

interface PersistedEvaluationInput {
  evaluationId: string;
  conversationId: string;
  provider: EvaluationProvider;
  model: string;
  scope: EvaluationScope;
  promptId?: string | null;
  promptName: string;
  promptText: string;
  selectionInstruction?: string | null;
  createdAt: string;
}

function providerForProjectPath(projectPath: string): "claude" | "codex" {
  return projectPath.startsWith("codex:") ? "codex" : "claude";
}

function conversationIdFor(projectPath: string, sessionId: string): string {
  return `${providerForProjectPath(projectPath)}:${sessionId}`;
}

function openDb() {
  ensureDatabaseReady();
  const db = new Database(OLTP_DB_PATH);
  db.pragma("journal_mode = WAL");
  db.pragma("foreign_keys = ON");
  db.pragma("busy_timeout = 5000");
  return db;
}

function ensureDatabaseReady() {
  if (databaseReady) {
    return;
  }

  execFileSync(
    "uv",
    ["run", "alembic", "-c", join(process.cwd(), "alembic.ini"), "-x", "target=oltp", "upgrade", "head"],
    {
      cwd: process.cwd(),
      stdio: "pipe",
    }
  );

  const db = new Database(OLTP_DB_PATH);
  db.pragma("journal_mode = WAL");
  db.pragma("foreign_keys = ON");
  db.pragma("busy_timeout = 5000");
  seedDefaultPrompt(db);
  db.close();
  databaseReady = true;
}

function seedDefaultPrompt(db: Database.Database) {
  const now = new Date().toISOString();
  db.prepare(
    `
      INSERT OR IGNORE INTO evaluation_prompts (
        prompt_id,
        name,
        description,
        prompt_text,
        is_default,
        created_at,
        updated_at
      ) VALUES (?, ?, ?, ?, 1, ?, ?)
    `
  ).run(
    "default-conversation-review",
    DEFAULT_EVALUATION_PROMPT_NAME,
    "Default review prompt for diagnosing instruction quality and conversation flow.",
    DEFAULT_EVALUATION_PROMPT,
    now,
    now
  );
}

function mapPrompt(row: PromptRow): EvaluationPrompt {
  return {
    promptId: row.prompt_id,
    name: row.name,
    description: row.description,
    promptText: row.prompt_text,
    isDefault: Boolean(row.is_default),
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

function mapEvaluation(row: EvaluationRow): ConversationEvaluation {
  return {
    evaluationId: row.evaluation_id,
    conversationId: row.conversation_id,
    promptId: row.prompt_id,
    provider: row.provider,
    model: row.model,
    status: row.status,
    scope: row.scope,
    selectionInstruction: row.selection_instruction,
    promptName: row.prompt_name,
    promptText: row.prompt_text,
    reportMarkdown: row.report_markdown,
    rawOutput: row.raw_output,
    errorMessage: row.error_message,
    command: row.command,
    createdAt: row.created_at,
    finishedAt: row.finished_at,
    durationMs: row.duration_ms,
  };
}

function buildEvaluationCommand(
  provider: EvaluationProvider,
  model: string,
  cwd: string
): { binary: string; args: string[]; command: string } {
  if (provider === "claude") {
    const args = ["-p", "--dangerously-skip-permissions", "--model", model];
    return {
      binary: "claude",
      args,
      command: ["claude", ...args].join(" "),
    };
  }

  const args = [
    "exec",
    "--dangerously-bypass-approvals-and-sandbox",
    "--skip-git-repo-check",
    "-C",
    cwd,
    "-m",
    model,
    "-",
  ];
  return {
    binary: "codex",
    args,
    command: ["codex", ...args].join(" "),
  };
}

function decodeWorkspacePath(projectPath: string): string {
  const encoded = projectPath.startsWith("codex:")
    ? projectPath.slice("codex:".length)
    : projectPath;

  if (encoded.startsWith("-")) {
    return encoded.replace(/^-/, "/").replace(/-/g, "/");
  }

  return join(homedir(), encoded);
}

function resolveWorkspace(projectPath: string): string {
  const candidate = decodeWorkspacePath(projectPath);
  if (existsSync(candidate)) {
    return candidate;
  }
  return process.cwd();
}

function formatBlock(messageId: string, block: ProcessedConversation["messages"][number]["blocks"][number]): string {
  if (block.type === "text") {
    return `message ${messageId} text:\n${block.text}`;
  }
  if (block.type === "thinking") {
    return `message ${messageId} thinking:\n${block.thinking}`;
  }

  return [
    `message ${messageId} tool call: ${block.toolName}`,
    `input: ${JSON.stringify(block.input, null, 2)}`,
    block.result ? `result:\n${block.result}` : null,
    block.isError ? "status: failed" : "status: succeeded",
  ]
    .filter(Boolean)
    .join("\n");
}

function renderConversationTranscript(
  conversation: ProcessedConversation,
  scope: EvaluationScope
): string {
  if (scope === "failed_tool_calls") {
    const segments = conversation.messages.flatMap((message, index) => {
      const hasFailure = message.blocks.some(
        (block) => block.type === "tool_call" && block.isError
      );
      if (!hasFailure) {
        return [];
      }

      const contextMessages = [conversation.messages[index - 1], message].filter(Boolean);
      return contextMessages.map((entry) => {
        const blocks = entry.blocks.map((block) => formatBlock(entry.id, block)).join("\n\n");
        return [
          `message_id: ${entry.id}`,
          `role: ${entry.role}`,
          `timestamp: ${new Date(entry.timestamp).toISOString()}`,
          blocks,
        ].join("\n");
      });
    });

    if (segments.length === 0) {
      return "No failed tool calls were captured for this conversation.";
    }

    return segments.join("\n\n---\n\n");
  }

  return conversation.messages
    .map((message) => {
      const blocks = message.blocks.map((block) => formatBlock(message.id, block)).join("\n\n");
      return [
        `message_id: ${message.id}`,
        `role: ${message.role}`,
        `timestamp: ${new Date(message.timestamp).toISOString()}`,
        blocks,
      ].join("\n");
    })
    .join("\n\n---\n\n");
}

function buildEvaluationPrompt(
  conversation: ProcessedConversation,
  scope: EvaluationScope,
  promptText: string,
  selectionInstruction?: string | null
): string {
  const scopeLabel =
    scope === "full"
      ? "the full conversation"
      : scope === "failed_tool_calls"
        ? "only failed tool calls and their nearby context"
        : "a guided subset of the conversation";

  const selectionBlock =
    scope === "guided_subset" && selectionInstruction
      ? `Before writing the report, first isolate only the messages that match this analyst instruction:\n${selectionInstruction}\n\nThen base the report on that subset.`
      : "";

  return [
    "You are evaluating a coding-assistant conversation for quality, instruction design, and flow.",
    "Do not edit files or run commands. Analyze only the provided transcript.",
    `The transcript below contains ${scopeLabel}.`,
    promptText.trim(),
    selectionBlock,
    "Conversation metadata:",
    `- model: ${conversation.model ?? "unknown"}`,
    `- message count: ${conversation.messages.length}`,
    `- sub-agent count: ${conversation.subagents.length}`,
    `- started at: ${new Date(conversation.startTime).toISOString()}`,
    "",
    "Transcript:",
    renderConversationTranscript(conversation, scope),
  ]
    .filter(Boolean)
    .join("\n\n");
}

async function executeEvaluationCommand(
  provider: EvaluationProvider,
  model: string,
  prompt: string,
  cwd: string
): Promise<{ stdout: string; stderr: string; command: string }> {
  const { binary, args, command } = buildEvaluationCommand(provider, model, cwd);

  return new Promise((resolve, reject) => {
    const child = spawn(binary, args, {
      cwd,
      stdio: "pipe",
      env: {
        ...process.env,
        CI: "1",
      },
    });

    let stdout = "";
    let stderr = "";

    child.stdout.on("data", (chunk) => {
      stdout += chunk.toString();
    });

    child.stderr.on("data", (chunk) => {
      stderr += chunk.toString();
    });

    child.on("error", reject);
    child.on("close", (code) => {
      if (code === 0) {
        resolve({
          stdout: stdout.trim(),
          stderr: stderr.trim(),
          command,
        });
        return;
      }

      reject(
        new Error(
          stderr.trim() ||
            stdout.trim() ||
            `${binary} exited with status ${code ?? "unknown"}.`
        )
      );
    });

    child.stdin.write(prompt);
    child.stdin.end();
  });
}

function insertRunningEvaluation(
  db: Database.Database,
  input: PersistedEvaluationInput,
  command: string
) {
  db.prepare(
    `
      INSERT INTO conversation_evaluations (
        evaluation_id,
        conversation_id,
        prompt_id,
        provider,
        model,
        status,
        scope,
        selection_instruction,
        prompt_name,
        prompt_text,
        report_markdown,
        raw_output,
        error_message,
        command,
        created_at,
        finished_at,
        duration_ms
      ) VALUES (?, ?, ?, ?, ?, 'running', ?, ?, ?, ?, NULL, NULL, NULL, ?, ?, NULL, NULL)
    `
  ).run(
    input.evaluationId,
    input.conversationId,
    input.promptId ?? null,
    input.provider,
    input.model,
    input.scope,
    input.selectionInstruction ?? null,
    input.promptName,
    input.promptText,
    command,
    input.createdAt
  );
}

function getEvaluationById(db: Database.Database, evaluationId: string) {
  return mapEvaluation(
    db.prepare(
      `
        SELECT
          evaluation_id,
          conversation_id,
          prompt_id,
          provider,
          model,
          status,
          scope,
          selection_instruction,
          prompt_name,
          prompt_text,
          report_markdown,
          raw_output,
          error_message,
          command,
          created_at,
          finished_at,
          duration_ms
        FROM conversation_evaluations
        WHERE evaluation_id = ?
      `
    ).get(evaluationId) as EvaluationRow
  );
}

async function runConversationEvaluationJob(args: {
  evaluationId: string;
  provider: EvaluationProvider;
  model: string;
  prompt: string;
  workspace: string;
  fallbackCommand: string;
}) {
  const startedAt = Date.now();

  try {
    const result = await executeEvaluationCommand(
      args.provider,
      args.model,
      args.prompt,
      args.workspace
    );
    const db = openDb();
    try {
      db.prepare(
        `
          UPDATE conversation_evaluations
          SET
            status = 'completed',
            report_markdown = ?,
            raw_output = ?,
            command = ?,
            finished_at = ?,
            duration_ms = ?
          WHERE evaluation_id = ?
        `
      ).run(
        result.stdout,
        result.stdout,
        result.command,
        new Date().toISOString(),
        Date.now() - startedAt,
        args.evaluationId
      );
    } finally {
      db.close();
    }
  } catch (error) {
    const message = error instanceof Error ? error.message : "Evaluation failed.";
    const db = openDb();
    try {
      db.prepare(
        `
          UPDATE conversation_evaluations
          SET
            status = 'failed',
            error_message = ?,
            command = ?,
            finished_at = ?,
            duration_ms = ?
          WHERE evaluation_id = ?
        `
      ).run(
        message,
        args.fallbackCommand,
        new Date().toISOString(),
        Date.now() - startedAt,
        args.evaluationId
      );
    } finally {
      db.close();
    }
  }
}

export function listEvaluationPrompts(): EvaluationPrompt[] {
  const db = openDb();
  try {
    const rows = db
      .prepare(
        `
          SELECT prompt_id, name, description, prompt_text, is_default, created_at, updated_at
          FROM evaluation_prompts
          ORDER BY is_default DESC, updated_at DESC, name ASC
        `
      )
      .all() as PromptRow[];
    return rows.map(mapPrompt);
  } finally {
    db.close();
  }
}

export function createEvaluationPrompt(input: {
  name: string;
  description?: string | null;
  promptText: string;
}): EvaluationPrompt {
  const db = openDb();
  const now = new Date().toISOString();
  const promptId = randomUUID();

  try {
    db.prepare(
      `
        INSERT INTO evaluation_prompts (
          prompt_id,
          name,
          description,
          prompt_text,
          is_default,
          created_at,
          updated_at
        ) VALUES (?, ?, ?, ?, 0, ?, ?)
      `
    ).run(promptId, input.name, input.description ?? null, input.promptText, now, now);

    return mapPrompt(
      db.prepare(
        `
          SELECT prompt_id, name, description, prompt_text, is_default, created_at, updated_at
          FROM evaluation_prompts
          WHERE prompt_id = ?
        `
      ).get(promptId) as PromptRow
    );
  } finally {
    db.close();
  }
}

export function updateEvaluationPrompt(
  promptId: string,
  input: {
    name: string;
    description?: string | null;
    promptText: string;
  }
): EvaluationPrompt {
  const db = openDb();
  const now = new Date().toISOString();

  try {
    db.prepare(
      `
        UPDATE evaluation_prompts
        SET name = ?, description = ?, prompt_text = ?, updated_at = ?
        WHERE prompt_id = ?
      `
    ).run(input.name, input.description ?? null, input.promptText, now, promptId);

    return mapPrompt(
      db.prepare(
        `
          SELECT prompt_id, name, description, prompt_text, is_default, created_at, updated_at
          FROM evaluation_prompts
          WHERE prompt_id = ?
        `
      ).get(promptId) as PromptRow
    );
  } finally {
    db.close();
  }
}

export function deleteEvaluationPrompt(promptId: string) {
  const db = openDb();

  try {
    const row = db
      .prepare("SELECT is_default FROM evaluation_prompts WHERE prompt_id = ?")
      .get(promptId) as { is_default: number } | undefined;

    if (!row) {
      throw new Error("Prompt not found.");
    }

    if (row.is_default) {
      throw new Error("The default prompt cannot be deleted.");
    }

    db.prepare("DELETE FROM evaluation_prompts WHERE prompt_id = ?").run(promptId);
  } finally {
    db.close();
  }
}

export function listConversationEvaluations(conversationId: string): ConversationEvaluation[] {
  const db = openDb();
  try {
    const rows = db
      .prepare(
        `
          SELECT
            evaluation_id,
            conversation_id,
            prompt_id,
            provider,
            model,
            status,
            scope,
            selection_instruction,
            prompt_name,
            prompt_text,
            report_markdown,
            raw_output,
            error_message,
            command,
            created_at,
            finished_at,
            duration_ms
          FROM conversation_evaluations
          WHERE conversation_id = ?
          ORDER BY created_at DESC
        `
      )
      .all(conversationId) as EvaluationRow[];

    return rows.map(mapEvaluation);
  } finally {
    db.close();
  }
}

export async function createConversationEvaluation(
  input: EvaluationRequest
): Promise<ConversationEvaluation> {
  const conversation = await getConversation(input.projectPath, input.sessionId);
  if (!conversation) {
    throw new Error("Conversation not found.");
  }

  const conversationId = conversationIdFor(input.projectPath, input.sessionId);
  const promptName = (input.promptName || DEFAULT_EVALUATION_PROMPT_NAME).trim();
  const promptText = input.promptText.trim();
  if (!promptName || !promptText) {
    throw new Error("Prompt name and prompt text are required.");
  }

  const db = openDb();
  const evaluationId = randomUUID();
  const createdAt = new Date().toISOString();
  const workspace = resolveWorkspace(input.projectPath);
  const fullPrompt = buildEvaluationPrompt(
    conversation,
    input.scope,
    promptText,
    input.selectionInstruction
  );
  const { command } = buildEvaluationCommand(input.provider, input.model, workspace);

  let evaluation: ConversationEvaluation;
  try {
    insertRunningEvaluation(
      db,
      {
        evaluationId,
        conversationId,
        promptId: input.promptId ?? null,
        provider: input.provider,
        model: input.model,
        scope: input.scope,
        selectionInstruction: input.selectionInstruction ?? null,
        promptName,
        promptText,
        createdAt,
      },
      command
    );
    evaluation = getEvaluationById(db, evaluationId);
  } finally {
    db.close();
  }

  void runConversationEvaluationJob({
    evaluationId,
    provider: input.provider,
    model: input.model,
    prompt: fullPrompt,
    workspace,
    fallbackCommand: command,
  });

  return evaluation;
}

File: python/helaicopter_api/application/evaluations.py
"""Application-layer conversation evaluation job orchestration."""

from __future__ import annotations

import json
from datetime import UTC, datetime
from pathlib import Path

from pydantic import ConfigDict, InstanceOf, validate_call

from helaicopter_api.application.conversations import get_conversation
from helaicopter_api.application.evaluation_prompts import resolve_evaluation_prompt
from helaicopter_api.bootstrap.services import BackendServices
from helaicopter_api.ports.evaluations import EvaluationJobRequest, EvaluationJobResult
from helaicopter_api.schema.conversations import ConversationDetailResponse, ConversationMessageBlockResponse
from helaicopter_api.schema.evaluations import (
    ConversationEvaluationCreateRequest,
    ConversationEvaluationResponse,
)


class ConversationEvaluationConversationNotFoundError(LookupError):
    """Raised when the requested conversation cannot be loaded."""


@validate_call(config=ConfigDict(strict=True), validate_return=True)
def list_conversation_evaluations(
    services: InstanceOf[BackendServices],
    *,
    project_path: str,
    session_id: str,
) -> list[ConversationEvaluationResponse]:
    """Return persisted evaluations for one conversation, newest first."""
    conversation_id = _conversation_id_for(project_path, session_id)
    records = services.app_sqlite_store.list_conversation_evaluations(conversation_id)
    return [_to_response(record) for record in records]


@validate_call(config=ConfigDict(strict=True), validate_return=True)
def create_conversation_evaluation(
    services: InstanceOf[BackendServices],
    *,
    project_path: str,
    session_id: str,
    body: ConversationEvaluationCreateRequest,
) -> ConversationEvaluationResponse:
    """Create a persisted evaluation job and submit it to the runner."""
    if services.evaluation_job_runner is None:
        raise RuntimeError("Evaluation job runner is unavailable.")

    conversation = get_conversation(
        services,
        project_path=project_path,
        session_id=session_id,
    )
    if conversation is None:
        raise ConversationEvaluationConversationNotFoundError("Conversation not found.")

    prompt_id, prompt_name, prompt_text = _resolve_prompt_content(services, body)
    workspace = _resolve_workspace(services, project_path)
    job_request = EvaluationJobRequest(
        evaluation_id=f"pending-{session_id}",
        provider=body.provider,
        model=body.model,
        workspace=workspace,
        prompt=_build_evaluation_prompt(
            conversation,
            scope=body.scope,
            prompt_text=prompt_text,
            selection_instruction=body.selection_instruction,
        ),
    )
    command = services.evaluation_job_runner.describe_command(job_request)
    record = services.app_sqlite_store.create_conversation_evaluation(
        conversation_id=_conversation_id_for(project_path, session_id),
        provider=body.provider,
        model=body.model,
        status="running",
        scope=body.scope,
        prompt_id=prompt_id,
        prompt_name=prompt_name,
        prompt_text=prompt_text,
        selection_instruction=body.selection_instruction,
        command=command,
    )
    submitted_request = EvaluationJobRequest(
        evaluation_id=record.evaluation_id,
        provider=job_request.provider,
        model=job_request.model,
        workspace=job_request.workspace,
        prompt=job_request.prompt,
        timeout_seconds=job_request.timeout_seconds,
    )
    try:
        services.evaluation_job_runner.submit(
            submitted_request,
            lambda result: _persist_job_result(services, result),
        )
    except Exception as error:
        services.app_sqlite_store.update_conversation_evaluation(
            record.evaluation_id,
            status="failed",
            command=command,
            report_markdown=None,
            raw_output=None,
            error_message=str(error) or "Failed to submit evaluation job.",
            finished_at=_now_iso(),
            duration_ms=0,
        )
        raise RuntimeError("Failed to start evaluation job.") from error

    return _to_response(record)


def _persist_job_result(services: BackendServices, result: EvaluationJobResult) -> None:
    services.app_sqlite_store.update_conversation_evaluation(
        result.evaluation_id,
        status=result.status,
        command=result.command,
        report_markdown=result.report_markdown,
        raw_output=result.raw_output,
        error_message=result.error_message,
        finished_at=result.finished_at,
        duration_ms=result.duration_ms,
    )


def _resolve_prompt_content(
    services: BackendServices,
    body: ConversationEvaluationCreateRequest,
) -> tuple[str | None, str, str]:
    if body.prompt_id is not None:
        prompt = resolve_evaluation_prompt(services, prompt_id=body.prompt_id)
        return prompt.prompt_id, prompt.name, prompt.prompt_text

    if body.prompt_name and body.prompt_text:
        return None, body.prompt_name, body.prompt_text

    prompt = resolve_evaluation_prompt(services)
    return prompt.prompt_id, prompt.name, prompt.prompt_text


def _build_evaluation_prompt(
    conversation: ConversationDetailResponse,
    *,
    scope: str,
    prompt_text: str,
    selection_instruction: str | None,
) -> str:
    scope_label = {
        "full": "the full conversation",
        "failed_tool_calls": "only failed tool calls and their nearby context",
        "guided_subset": "a guided subset of the conversation",
    }[scope]
    selection_block = ""
    if scope == "guided_subset" and selection_instruction:
        selection_block = (
            "Before writing the report, first isolate only the messages that match this analyst instruction:\n"
            f"{selection_instruction}\n\nThen base the report on that subset."
        )
    return "\n\n".join(
        part
        for part in [
            "You are evaluating a coding-assistant conversation for quality, instruction design, and flow.",
            "Do not edit files or run commands. Analyze only the provided transcript.",
            f"The transcript below contains {scope_label}.",
            prompt_text.strip(),
            selection_block,
            "Conversation metadata:",
            f"- model: {conversation.model or 'unknown'}",
            f"- message count: {len(conversation.messages)}",
            f"- sub-agent count: {len(conversation.subagents)}",
            f"- started at: {_isoformat_epoch_ms(conversation.start_time)}",
            "",
            "Transcript:",
            _render_conversation_transcript(conversation, scope=scope),
        ]
        if part
    )


def _render_conversation_transcript(conversation: ConversationDetailResponse, *, scope: str) -> str:
    if scope == "failed_tool_calls":
        segments: list[str] = []
        for index, message in enumerate(conversation.messages):
            has_failure = any(block.type == "tool_call" and block.is_error for block in message.blocks)
            if not has_failure:
                continue
            for contextual_message in conversation.messages[max(0, index - 1) : index + 1]:
                segments.append(_render_message(contextual_message))
        if not segments:
            return "No failed tool calls were captured for this conversation."
        return "\n\n---\n\n".join(segments)

    return "\n\n---\n\n".join(_render_message(message) for message in conversation.messages)


def _render_message(message: object) -> str:
    blocks = getattr(message, "blocks", [])
    return "\n".join(
        [
            f"message_id: {getattr(message, 'id')}",
            f"role: {getattr(message, 'role')}",
            f"timestamp: {_isoformat_epoch_ms(getattr(message, 'timestamp'))}",
            "\n\n".join(_format_block(block) for block in blocks),
        ]
    )


def _format_block(block: ConversationMessageBlockResponse) -> str:
    if block.type == "text":
        return f"message text:\n{block.text or ''}"
    if block.type == "thinking":
        return f"message thinking:\n{block.thinking or ''}"

    tool_input = block.input.root
    parts = [
        f"message tool call: {block.tool_name or 'unknown'}",
        f"input: {json.dumps(tool_input, indent=2)}",
        f"result:\n{block.result}" if block.result else None,
        "status: failed" if block.is_error else "status: succeeded",
    ]
    return "\n".join(part for part in parts if part)


def _resolve_workspace(services: BackendServices, project_path: str) -> Path:
    """Resolve a filesystem workspace for the given `project_path`.

    Preference order:
    1) If the configured project root exists, prefer it. This keeps evaluation
       jobs scoped to the active backend workspace or test fixture.
    2) Otherwise, if the current working directory is inside the decoded
       project path, use the CWD. This supports running within a git worktree
       while the project path points at the repository root.
    3) Otherwise, if the decoded path exists on disk, use it.
    4) Fallback to the configured `settings.project_root`.
    """
    encoded = project_path
    if encoded.startswith("codex:"):
        encoded = encoded[len("codex:") :]
    if encoded.startswith("-"):
        decoded = Path("/" + encoded.lstrip("-").replace("-", "/"))
    else:
        decoded = Path.home() / encoded

    configured_root = services.settings.project_root.resolve()
    if configured_root.exists():
        return configured_root

    try:
        cwd = Path.cwd().resolve()
        decoded_resolved = decoded.resolve()
        # If running inside a worktree or subdirectory of the decoded path,
        # prefer the active working directory for better locality when there is
        # no explicit project_root override.
        if cwd.is_relative_to(decoded_resolved):
            return cwd
    except Exception:
        # If resolution fails for any reason, fall back to the next options.
        pass

    if decoded.exists():
        return decoded
    return services.settings.project_root


def _conversation_id_for(project_path: str, session_id: str) -> str:
    provider = "codex" if project_path.startswith("codex:") else "claude"
    return f"{provider}:{session_id}"


def _isoformat_epoch_ms(timestamp_ms: float) -> str:
    return datetime.fromtimestamp(timestamp_ms / 1000, tz=UTC).isoformat()


def _to_response(record: object) -> ConversationEvaluationResponse:
    if hasattr(record, "model_dump"):
        return ConversationEvaluationResponse.model_validate(record.model_dump(mode="python"))
    return ConversationEvaluationResponse.model_validate(record)


def _now_iso() -> str:
    return datetime.now(UTC).isoformat()
